File: APISwitcher/Services/BalanceService.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Nodes;
using APISwitcher.Models;

namespace APISwitcher.Services;

/// <summary>
/// 余额查询服务
/// </summary>
public class BalanceService : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly JsonSerializerOptions _jsonOptions;

    public BalanceService()
    {
        _httpClient = new HttpClient();
        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true
        };
    }

    /// <summary>
    /// 查询Profile的余额
    /// </summary>
    public async Task<BalanceInfo> QueryBalanceAsync(Profile profile)
    {
        // 检查是否配置了余额API（在最开始检查，避免创建不必要的对象）
        if (profile.BalanceApi == null)
        {
            throw new InvalidOperationException("未配置余额API");
        }

        // 获取当前的 BalanceInfo（如果存在），保留历史状态
        var balanceInfo = profile.BalanceInfo ?? new BalanceInfo { IsFirstLoad = true };
        balanceInfo.IsLoading = true;

        try
        {

            // 从Profile的Settings中获取环境变量
            var token = GetTokenFromSettings(profile.Settings);
            var baseUrl = GetBaseUrlFromSettings(profile.Settings);

            if (string.IsNullOrEmpty(token) || string.IsNullOrEmpty(baseUrl))
            {
                balanceInfo.HasError = true;
                balanceInfo.ErrorMessage = "缺少认证信息";
                balanceInfo.IsLoading = false;
                return balanceInfo;
            }

            // 构建完整URL
            var fullUrl = CombineUrl(baseUrl, profile.BalanceApi.Endpoint);

            // 创建请求
            using var request = CreateHttpRequest(profile.BalanceApi, token, fullUrl);

            // 设置超时
            using var cts = new CancellationTokenSource(profile.BalanceApi.Timeout);

            // 发送请求
            var response = await _httpClient.SendAsync(request, cts.Token);
            response.EnsureSuccessStatusCode();

            // 解析响应
            var responseBody = await response.Content.ReadAsStringAsync();
            var jsonNode = JsonNode.Parse(responseBody);

            if (jsonNode == null)
            {
                throw new Exception("响应内容为空");
            }

            // 计算余额（返回 long）
            var rawBalance = CalculateBalance(jsonNode, profile.BalanceApi);
            decimal balance = rawBalance;

            // 应用除数进行单位转换（使用 decimal 保证精度）
            if (profile.BalanceApi.Divisor > 0 && profile.BalanceApi.Divisor != 1.0)
            {
                balance = balance / (decimal)profile.BalanceApi.Divisor;
            }

            var isUnlimited = false;

            if (!string.IsNullOrEmpty(profile.BalanceApi.UnlimitedField))
            {
                try
                {
                    isUnlimited = ExtractValue<bool>(jsonNode, profile.BalanceApi.UnlimitedField);
                }
                catch
                {
                    // 如果提取失败，默认为false
                    isUnlimited = false;
                }
            }

            // 查询成功，更新余额信息
            balanceInfo.Balance = balance;
            balanceInfo.IsUnlimited = isUnlimited;
            balanceInfo.DisplayUnit = profile.BalanceApi.DisplayUnit;
            balanceInfo.HasError = false;
            balanceInfo.FailureCount = 0;  // 重置失败计数
            balanceInfo.LastSuccessBalance = balance;  // 保存成功的余额
            balanceInfo.IsFirstLoad = false;  // 标记已完成首次加载
        }
        catch (OperationCanceledException)
        {
            balanceInfo.HasError = true;
            balanceInfo.ErrorMessage = "请求超时";
            balanceInfo.FailureCount++;  // 增加失败计数
        }
        catch (HttpRequestException ex)
        {
            balanceInfo.HasError = true;
            balanceInfo.ErrorMessage = $"网络错误: {ex.Message}";
            balanceInfo.FailureCount++;  // 增加失败计数
        }
        catch (Exception ex)
        {
            balanceInfo.HasError = true;
            balanceInfo.ErrorMessage = $"查询失败: {ex.Message}";
            balanceInfo.FailureCount++;  // 增加失败计数
        }
        finally
        {
            balanceInfo.IsLoading = false;
        }

        return balanceInfo;
    }

    /// <summary>
    /// 计算余额
    /// </summary>
    private long CalculateBalance(JsonNode jsonNode, BalanceApiConfig apiConfig)
    {
        // 优先使用 balanceField
        if (!string.IsNullOrEmpty(apiConfig.BalanceField))
        {
            return ExtractValue<long>(jsonNode, apiConfig.BalanceField);
        }

        // 如果没有 balanceField，则使用 limitField - usedField
        if (!string.IsNullOrEmpty(apiConfig.LimitField) && !string.IsNullOrEmpty(apiConfig.UsedField))
        {
            var limit = ExtractValue<long>(jsonNode, apiConfig.LimitField);
            var used = ExtractValue<long>(jsonNode, apiConfig.UsedField);
            return limit - used;
        }

        throw new Exception("未配置有效的余额字段（balanceField 或 limitField+usedField）");
    }

    /// <summary>
    /// 从Settings中提取Token
    /// </summary>
    private string? GetTokenFromSettings(ClaudeSettings settings)
    {
        if (settings.ExtensionData?.TryGetValue("env", out var envElement) == true)
        {
            if (envElement.ValueKind == JsonValueKind.Object)
            {
                if (envElement.TryGetProperty("ANTHROPIC_AUTH_TOKEN", out var tokenElement))
                {
                    return tokenElement.GetString();
                }
            }
        }
        return null;
    }

    /// <summary>
    /// 从Settings中提取BaseUrl
    /// </summary>
    private string? GetBaseUrlFromSettings(ClaudeSettings settings)
    {
        if (settings.ExtensionData?.TryGetValue("env", out var envElement) == true)
        {
            if (envElement.ValueKind == JsonValueKind.Object)
            {
                if (envElement.TryGetProperty("ANTHROPIC_BASE_URL", out var baseUrlElement))
                {
                    return baseUrlElement.GetString();
                }
            }
        }
        return null;
    }

    /// <summary>
    /// 组合URL
    /// </summary>
    private string CombineUrl(string baseUrl, string endpoint)
    {
        baseUrl = baseUrl.TrimEnd('/');
        endpoint = endpoint.TrimStart('/');
        return $"{baseUrl}/{endpoint}";
    }

    /// <summary>
    /// 创建HTTP请求
    /// </summary>
    private HttpRequestMessage CreateHttpRequest(BalanceApiConfig apiConfig, string token, string url)
    {
        var method = apiConfig.Method.ToUpper() == "POST" ? HttpMethod.Post : HttpMethod.Get;
        var request = new HttpRequestMessage(method, url);

        if (apiConfig.AuthType.ToLower() == "header")
        {
            // 使用Header认证：Authorization: Bearer {token}
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        }
        else if (apiConfig.AuthType.ToLower() == "body")
        {
            // 使用Body认证：{"token": "{token}"}
            var bodyContent = new { token };
            request.Content = JsonContent.Create(bodyContent);
        }

        return request;
    }

    /// <summary>
    /// 从JSON中提取指定路径的值
    /// </summary>
    private T ExtractValue<T>(JsonNode jsonNode, string path)
    {
        var parts = path.Split('.');
        JsonNode? current = jsonNode;

        foreach (var part in parts)
        {
            if (current is JsonObject obj)
            {
                current = obj[part];
            }
            else
            {
                throw new Exception($"无法找到路径: {path}");
            }
        }

        if (current == null)
        {
            throw new Exception($"路径 {path} 的值为null");
        }

        // 根据类型转换
        if (typeof(T) == typeof(long))
        {
            return (T)(object)current.GetValue<long>();
        }
        else if (typeof(T) == typeof(bool))
        {
            return (T)(object)current.GetValue<bool>();
        }
        else if (typeof(T) == typeof(string))
        {
            return (T)(object)(current.GetValue<string>() ?? string.Empty);
        }

        throw new NotSupportedException($"不支持的类型: {typeof(T)}");
    }

    public void Dispose()
    {
        _httpClient?.Dispose();
    }
}

File: APISwitcher.macOS/APISwitcher/Services/BalanceService.swift
//
//  BalanceService.swift
//  APISwitcher
//
//  余额查询服务
//

import Foundation

final class BalanceService: Sendable {
    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    /// 查询余额
    func queryBalance(config: Profile) async throws -> BalanceInfo {
        guard let balanceApi = config.balanceApi else {
            throw BalanceError.notConfigured
        }

        // 从 settings 中提取 token 和 baseUrl
        let token = getTokenFromSettings(config.settings)
        let baseUrl = getBaseUrlFromSettings(config.settings)

        // 检查认证信息（cookie 认证不需要 token）
        if balanceApi.authType.lowercased() != "cookie" {
            guard !token.isEmpty else {
                throw BalanceError.missingToken
            }
        }

        guard !baseUrl.isEmpty else {
            throw BalanceError.missingBaseUrl
        }

        // 构建完整 URL
        let fullUrl = try combineUrl(baseUrl: baseUrl, endpoint: balanceApi.endpoint)

        // 创建请求
        var request = createHttpRequest(
            apiConfig: balanceApi,
            token: token,
            url: fullUrl
        )

        // 设置超时
        request.timeoutInterval = TimeInterval(balanceApi.timeout) / 1000.0

        // 发送请求
        let (data, response) = try await session.data(for: request)

        // 检查响应状态
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw BalanceError.requestFailed
        }

        // 解析 JSON
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw BalanceError.invalidResponse
        }

        // 计算余额
        var balance = try calculateBalance(from: json, apiConfig: balanceApi)

        // 应用除数进行单位转换
        if balanceApi.divisor > 0 && balanceApi.divisor != 1.0 {
            balance = balance / balanceApi.divisor
        }

        // 检查是否无限额度
        var isUnlimited = false
        if let unlimitedField = balanceApi.unlimitedField {
            isUnlimited = (try? extractValue(from: json, path: unlimitedField) as? Bool) ?? false
        }

        return BalanceInfo(
            balance: balance,
            currency: balanceApi.displayUnit,
            isUnlimited: isUnlimited
        )
    }

    /// 计算余额
    private func calculateBalance(from json: [String: Any], apiConfig: BalanceApiConfig) throws -> Double {
        // 优先使用 balanceField
        if let balanceField = apiConfig.balanceField, !balanceField.isEmpty {
            if let value = try? extractValue(from: json, path: balanceField) {
                if let doubleValue = value as? Double {
                    return doubleValue
                } else if let intValue = value as? Int {
                    return Double(intValue)
                } else if let stringValue = value as? String, let doubleValue = Double(stringValue) {
                    return doubleValue
                }
            }
        }

        // 如果没有 balanceField，则使用 limitField - usedField
        if let limitField = apiConfig.limitField, !limitField.isEmpty,
           let usedField = apiConfig.usedField, !usedField.isEmpty {
            if let limitValue = try? extractValue(from: json, path: limitField),
               let usedValue = try? extractValue(from: json, path: usedField) {
                let limit = (limitValue as? Double) ?? Double((limitValue as? Int) ?? 0)
                let used = (usedValue as? Double) ?? Double((usedValue as? Int) ?? 0)
                return limit - used
            }
        }

        throw BalanceError.balanceNotFound
    }

    /// 从 Settings 中提取 Token
    private func getTokenFromSettings(_ settings: ClaudeSettings) -> String {
        let dict = settings.toDictionary()
        if let env = dict["env"] as? [String: Any],
           let token = env["ANTHROPIC_AUTH_TOKEN"] as? String {
            return token
        }
        return ""
    }

    /// 从 Settings 中提取 BaseUrl
    private func getBaseUrlFromSettings(_ settings: ClaudeSettings) -> String {
        let dict = settings.toDictionary()
        if let env = dict["env"] as? [String: Any],
           let baseUrl = env["ANTHROPIC_BASE_URL"] as? String {
            return baseUrl
        }
        return ""
    }

    /// 组合 URL
    private func combineUrl(baseUrl: String, endpoint: String) throws -> URL {
        let cleanBase = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let cleanEndpoint = endpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let urlString = "\(cleanBase)/\(cleanEndpoint)"

        guard let url = URL(string: urlString) else {
            throw BalanceError.invalidURL
        }
        return url
    }

    /// 创建 HTTP 请求
    private func createHttpRequest(apiConfig: BalanceApiConfig, token: String, url: URL) -> URLRequest {
        let method = apiConfig.method.uppercased()
        var request = URLRequest(url: url)
        request.httpMethod = method

        if apiConfig.authType.lowercased() == "header" {
            // 使用 Header 认证：Authorization: Bearer {token}
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else if apiConfig.authType.lowercased() == "body" {
            // 使用 Body 认证：{"token": "{token}"}
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            let body = ["token": token]
            request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        } else if apiConfig.authType.lowercased() == "cookie" {
            // 使用 Cookie 认证
            if let sessionCookie = apiConfig.sessionCookie, !sessionCookie.isEmpty {
                request.setValue(sessionCookie, forHTTPHeaderField: "Cookie")
            }

            // 添加额外的 headers
            if let extraHeaders = apiConfig.extraHeaders {
                for (key, value) in extraHeaders {
                    request.setValue(value, forHTTPHeaderField: key)
                }
            }
        }

        return request
    }

    /// 从 JSON 中提取指定路径的值
    private func extractValue(from json: [String: Any], path: String) throws -> Any {
        let parts = path.split(separator: ".")
        var current: Any = json

        for part in parts {
            if let dict = current as? [String: Any],
               let value = dict[String(part)] {
                current = value
            } else {
                throw BalanceError.pathNotFound(path)
            }
        }

        return current
    }
}

enum BalanceError: LocalizedError {
    case notConfigured
    case missingToken
    case missingBaseUrl
    case invalidURL
    case requestFailed
    case invalidResponse
    case balanceNotFound
    case pathNotFound(String)

    var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "未配置余额 API"
        case .missingToken:
            return "缺少认证 Token"
        case .missingBaseUrl:
            return "缺少 Base URL"
        case .invalidURL:
            return "无效的 API 地址"
        case .requestFailed:
            return "请求失败"
        case .invalidResponse:
            return "无效的响应数据"
        case .balanceNotFound:
            return "无法从响应中提取余额"
        case .pathNotFound(let path):
            return "无法找到路径: \(path)"
        }
    }
}
